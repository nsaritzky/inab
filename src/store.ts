import { createStore } from "solid-js/store"
import type { Store, Transaction, Month, MonthYear, Panel } from "./types"
import { v4 as uuid } from "uuid"
import { createEffect, createMemo } from "solid-js"
import { dateToMonthYear } from "./utilities"

const sampleTxns: Transaction[] = [
  {
    id: "0",
    description: "test",
    amount: -1,
    payee: "BuyMart",
    envelope: "Groceries",
    account: "Checking",
    date: new Date(),
  },
  {
    id: "1",
    description: "another test",
    amount: -2,
    payee: "Scrooge McDuck",
    envelope: "Rent",
    account: "Checking",
    date: new Date(),
  },
]
export const initialState: Store = {
  transactions: [],
  accounts: [],
  currentMonth: "JUN 2023",
  unallocated: 0,
  envelopes: {
    Rent: { allocated: { "JUN 2023": 1000 } },
    Groceries: { allocated: { "JUN 2023": 300 } },
  },
  panel: "transactions",
}

export const createCentralStore = () => {
  const [state, setState] = createStore(initialState)

  const addTransaction = (
    idFn: () => string,
    {
      amount,
      date,
      payee,
      envelope,
      account,
      description,
    }: Omit<Transaction, "id">
  ) => {
    if (amount > 0) {
      setState("unallocated", (old) => old + amount)
    }
    setState("transactions", (txns) => [
      ...txns,
      { id: idFn(), amount, date, payee, envelope, account, description },
    ])
    if (!Object.keys(state.envelopes).includes(envelope)) {
      setState("envelopes", envelope, { allocated: {} })
    }
  }

  const deleteTransaction = (id: string) =>
    setState("transactions", (txns) => txns.filter((t) => t.id != id))

  const envelopeBalances = createMemo(() => {
    const result: Record<string, number> = {}
    for (const nm of Object.keys(state.envelopes)) {
      const activity = state.transactions
        .filter(
          (txn) =>
            txn.envelope == nm &&
            dateToMonthYear(txn.date) == state.currentMonth
        )
        .reduce((sum, txn) => sum + txn.amount, 0)
      Object.assign(result, { [nm]: activity })
    }
    return result
  })

  const accountBalances = createMemo(() => {
    const result: Record<string, number> = {}
    for (const nm of state.accounts) {
      const spent = state.transactions
        .filter((txn) => txn.account == nm)
        .reduce((sum, txn) => sum + txn.amount, 0)
      Object.assign(result, { [nm]: spent })
    }
    return result
  })

  const nextMonth = (month: Month): Month => {
    switch (month) {
      case "JAN":
        return "FEB"
      case "FEB":
        return "MAR"
      case "MAR":
        return "APR"
      case "APR":
        return "MAY"
      case "MAY":
        return "JUN"
      case "JUN":
        return "JUL"
      case "JUL":
        return "AUG"
      case "AUG":
        return "SEP"
      case "SEP":
        return "OCT"
      case "OCT":
        return "NOV"
      case "NOV":
        return "DEC"
      case "DEC":
        return "JAN"
    }
  }

  const prevMonth = (month: Month): Month => {
    switch (month) {
      case "JAN":
        return "DEC"
      case "FEB":
        return "JAN"
      case "MAR":
        return "FEB"
      case "APR":
        return "MAR"
      case "MAY":
        return "APR"
      case "JUN":
        return "MAY"
      case "JUL":
        return "JUN"
      case "AUG":
        return "JUL"
      case "SEP":
        return "AUG"
      case "OCT":
        return "SEP"
      case "NOV":
        return "OCT"
      case "DEC":
        return "NOV"
    }
  }

  const incMonth = (my: MonthYear): MonthYear => {
    const month = my.slice(0, 3) as Month
    const year = parseInt(my.slice(-4))
    let newYear
    if (month == "DEC") {
      newYear = year + 1
    } else {
      newYear = year
    }
    return `${nextMonth(month)} ${newYear}`
  }

  const decMonth = (my: MonthYear): MonthYear => {
    const month = my.slice(0, 3) as Month
    const year = parseInt(my.slice(-4))
    let newYear
    if (month == "JAN") {
      newYear = year - 1
    } else {
      newYear = year
    }
    return `${prevMonth(month)} ${newYear}`
  }

  const setIncMonth = () => setState("currentMonth", incMonth)
  const setDecMonth = () => setState("currentMonth", decMonth)

  const setPanel = (panel: Panel) => setState("panel", panel)

  return [
    state,
    {
      addTransaction,
      deleteTransaction,
      envelopeBalances,
      accountBalances,
      setIncMonth,
      setDecMonth,
      setPanel,
    },
  ] as const
}
