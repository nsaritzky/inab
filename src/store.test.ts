import { afterAll, beforeEach, describe, expect, test, vi } from "vitest"
import type { Transaction } from "./types"
import * as store from "./store"
import { createStore } from "solid-js/store"
import { createRoot } from "solid-js"

const sampleTxns = [
  {
    description: "test",
    amount: -1,
    payee: "BuyMart",
    envelope: "Groceries",
    account: "Checking",
    date: new Date(Date.parse("2023-06-28")),
  },
  {
    description: "another test",
    amount: -2,
    payee: "Scrooge McDuck",
    envelope: "Rent",
    account: "Checking",
    date: new Date(Date.parse("2023-06-30")),
  },
  {
    description: "another another test",
    amount: -10,
    payee: "Mr. F",
    envelope: "New",
    account: "Checking",
    date: new Date(Date.parse("2023-06-23")),
  },
  {
    description: "credit",
    amount: 10,
    payee: "",
    envelope: "",
    account: "Checking",
    date: new Date(Date.parse("2023-06-25")),
  },
]

const [dispose, state, fns] = createRoot((dispose) => {
  const [state, fns] = store.createCentralStore()
  return [dispose, state, fns]
})

let id = -1

const idFn = () => {
  id++
  return id.toString()
}

afterAll(dispose)

test("addTransaction", async () => {
  for (const txn of sampleTxns) {
    fns.addTransaction(idFn, txn)
  }

  await new Promise((done) => setTimeout(done, 0))
  // expect(state.transactions.length).toBe(4)
  // expect(Object.keys(state.envelopes)).toContain("New")
  // expect(state.envelopes["New"].allocated).toStrictEqual({})
  // expect(state.unallocated).toBe(10)
  expect(state).toMatchSnapshot()
})

describe("balances", () => {
  test("envelopeBalances", async () => {
    expect(fns.envelopeBalances()["Groceries"]).toBe(-1)
    expect(fns.envelopeBalances()["New"]).toBe(-10)
  })

  test.todo("accountBalances", () => {})
})

test("increment and decrement month", async () => {
  fns.setIncMonth()
  expect(state.currentMonth).toBe("JUL 2023")
  fns.setIncMonth()
  fns.setIncMonth()
  fns.setIncMonth()
  fns.setIncMonth()
  fns.setIncMonth()
  fns.setIncMonth()
  expect(state.currentMonth).toBe("JAN 2024")
  fns.setDecMonth()
  expect(state.currentMonth).toBe("DEC 2023")
  fns.setDecMonth()
  expect(state.currentMonth).toBe("NOV 2023")
})

test("deleteTransaction", async () => {
  fns.deleteTransaction("0")
  expect(state.transactions.length).toBe(3)
  expect(fns.envelopeBalances()["Rent"]).toBe(0)
})
