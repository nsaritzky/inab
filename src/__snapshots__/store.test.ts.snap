// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`addTransaction 1`] = `
{
  "accounts": [],
  "currentMonth": "JUN 2023",
  "envelopes": {
    "": {
      "allocated": {},
    },
    "Groceries": {
      "allocated": {
        "JUN 2023": 300,
      },
    },
    "New": {
      "allocated": {},
    },
    "Rent": {
      "allocated": {
        "JUN 2023": 1000,
      },
    },
  },
  "panel": "transactions",
  "transactions": [
    {
      "account": "Checking",
      "amount": -1,
      "date": 2023-06-28T00:00:00.000Z,
      "description": "test",
      "envelope": "Groceries",
      "id": "0",
      "payee": "BuyMart",
    },
    {
      "account": "Checking",
      "amount": -2,
      "date": 2023-06-30T00:00:00.000Z,
      "description": "another test",
      "envelope": "Rent",
      "id": "1",
      "payee": "Scrooge McDuck",
    },
    {
      "account": "Checking",
      "amount": -10,
      "date": 2023-06-23T00:00:00.000Z,
      "description": "another another test",
      "envelope": "New",
      "id": "2",
      "payee": "Mr. F",
    },
    {
      "account": "Checking",
      "amount": 10,
      "date": 2023-06-25T00:00:00.000Z,
      "description": "credit",
      "envelope": "",
      "id": "3",
      "payee": "",
    },
  ],
  "unallocated": 10,
}
`;
